### ISSUE TITLE
Add option to disable 'HTTP/SNI parser'

### ISSUE THREAD

**Issue by PouuleT** [shadowsocks/shadowsocks-libev#1503]
### What version of shadowsocks-libev are you using?

Any versions > v2.5.0

Specifically it seems to be on all versions after #809

### What operating system are you using?

Server : Debian
Gateway : OpenWRT
Client on local network : Debian

### What did you do?

Setup :
A ss-server on a distant server
A ss-redir client on the gateway of my local network

A specific DNS entry on a machine in my local network, example in /etc/hosts :
216.58.213.163 shadowsocks.org

( 216.58.208.238 -> IP of google.com at the moment of the test )

Action :

```
curl shadowsocks.org -vvv
```

### What did you expect to see?

I expected my request to go to the server resolved on my local host ( 216.58.208.238 in my case )

The curl should then give me an error page saying that the requested URL is not found on the server.
The same error you would see by doing this on a basic configuration :

============================================================================

```
curl -H 'Host: shadowsocks.org' 216.58.208.238 -vvv
* Rebuilt URL to: 216.58.208.238/
* Hostname was NOT found in DNS cache
*   Trying 216.58.208.238...
* Connected to 216.58.208.238 (216.58.208.238) port 80 (#0)
> GET / HTTP/1.1
> User-Agent: curl/7.38.0
> Accept: */*
> Host: shadowsocks.org
>
< HTTP/1.1 404 Not Found
< Content-Type: text/html; charset=UTF-8
< Referrer-Policy: no-referrer
< Content-Length: 1561
< Date: Fri, 19 May 2017 15:09:25 GMT
<
<!DOCTYPE html>
<html lang=en>
  <meta charset=utf-8>
  <meta name=viewport content="initial-scale=1, minimum-scale=1, width=device-width">
  <title>Error 404 (Not Found)!!1</title>
  <style>
    *{margin:0;padding:0}html,code{font:15px/22px arial,sans-serif}html{background:#fff;color:#222;padding:15px}body{margin:7% auto 0;max-width:390px;min-height:180px;padding:30px 0 15px}* > body{background:url(//www.google.com/images/errors/robot.png) 100% 5px no-repeat;padding-right:205px}p{margin:11px 0 22px;overflow:hidden}ins{color:#777;text-decoration:none}a img{border:0}@media screen and (max-width:772px){body{background:none;margin-top:0;max-width:none;padding-right:0}}#logo{background:url(//www.google.com/images/branding/googlelogo/1x/googlelogo_color_150x54dp.png) no-repeat;margin-left:-5px}@media only screen and (min-resolution:192dpi){#logo{background:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) no-repeat 0% 0%/100% 100%;-moz-border-image:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) 0}}@media only screen and (-webkit-min-device-pixel-ratio:2){#logo{background:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) no-repeat;-webkit-background-size:100% 100%}}#logo{display:inline-block;height:54px;width:150px}
  </style>
  <a href=//www.google.com/><span id=logo aria-label=Google></span></a>
  <p><b>404.</b> <ins>That’s an error.</ins>
  <p>The requested URL <code>/</code> was not found on this server.  <ins>That’s all we know.</ins>
* Connection #0 to host 216.58.208.238 left intact
```

============================================================================

### What did you see instead?

Instead, I got this :

============================================================================

```
curl shadowsocks.org -vvv
* Rebuilt URL to: shadowsocks.org/
*   Trying 216.58.208.238...
* Connected to shadowsocks.org (216.58.208.238) port 80 (#0)
> GET / HTTP/1.1
> User-Agent: curl/7.38.0
> Host: shadowsocks.org
> Accept: */*
>
< HTTP/1.1 200 OK
< Date: Fri, 19 May 2017 15:12:54 GMT
< Content-Type: text/html; charset=utf-8
< Transfer-Encoding: chunked
< Connection: keep-alive
< Last-Modified: Sun, 09 Apr 2017 01:45:12 GMT
< Access-Control-Allow-Origin: *
< Expires: Fri, 19 May 2017 15:22:54 GMT
< Cache-Control: max-age=600
< X-GitHub-Request-Id: CCCCCCCCCCCC
< X-Content-Type-Options: nosniff
* Server cloudflare-nginx is not blacklisted
< Server: cloudflare-nginx
< CF-RAY: XXXXXXXXXXX
<
<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta http-equiv="refresh" content="0; URL=/en/index.html">
<script type="text/javascript">
//<![CDATA[
try{if (!window.CloudFlare) {var CloudFlare=[{verbose:0,p:0,byc:0,owlid:"cf",bag2:1,mirage2:0,oracle:0,paths:{cloudflare:"/cdn-cgi/nexp/dok3v=XXXX/"},atok:"xxxxx",petok:"aaaaa",zone:"shadowsocks.org",rocket:"0",apps:{"abetterbrowser": {"ie": "6"}}}];!function(a,b){a=document.createElement("script"),b=document.getElementsByTagName("script")[0],a.async=!0,a.src="//ajax.cloudflare.com/cdn-cgi/nexp/dok3v=777777/cloudflare.min.js",b.parentNode.insertBefore(a,b)}()}}catch(e){};
//]]>
</script>
</head>
<body></body>
</html>
* Connection #0 to host shadowsocks.org left intact
```

-> This is the result of a curl on the real shadowsocks.org

Despite being resolved to a specific IP address, the server, before sending the request, did a DNS request, resolved shadowsocks.org to its real IP.

### What is your config in detail (with all sensitive info masked)?

ss-redir :
```
{
        "server": "11.22.33.44",
        "server_port": 1091,
        "local_port": 8381,
        "password": "XXXXXXXXXX",
        "timeout": 86400,
        "method": "aes-256-cfb",
}
```

ss-server :
```
{
    "server":"0.0.0.0",
    "server_port":1091,
    "local_port":8381,
    "password":"XXXXXXXXX",
    "timeout":86400,
    "method": "aes-256-cfb"
}
```

It seems to have been introduced by this pull request : https://github.com/shadowsocks/shadowsocks-libev/pull/809
More specifically this part about SNI ( https://github.com/shadowsocks/shadowsocks-libev/pull/809/files#diff-db5376f39c9e516b8a94af8d280d3d14R226 ) that parses the http packet to get the hostname.
And this part ( https://github.com/shadowsocks/shadowsocks-libev/pull/809/files#diff-db5376f39c9e516b8a94af8d280d3d14R417 ) that overrides the IP address the packet was destined to, and replace it with the hostname of the request
Am I correct?

I can understand that this is an appreciated feature when DNS requests are poisoned ( as stated here : https://github.com/shadowsocks/shadowsocks-libev/issues/944 )

But it's kind of annoying for users on the local network to see their requests being 'wrongly' redirected despite their own specific DNS configuration

But it would be nice if such a feature could be opted-out if needed don't you think? Do you think such an option could be added?

**Comment by madeye** [comment]
No plan for this. You can always avoid this kind of behavior with policy routing. More specifically, try dnsmasq and ipset.

**Comment by madeye** [comment]
BTW, putting domains into the bypass section in the ACL file would also do the trick.

**Comment by PouuleT** [comment]
ACL or policy routing might work it's true! But they would have to be specific to a domain or an IP.
It would be better if it was possible to specify that we want to always use the user's DNS resolution and not try to be more clever and resolve it on server's side

Does "No plan for this" means it will never appear in the project? Even if I manage to try a simple non-intrusive pull request?

**Comment by madeye** [comment]
Yes, you can open a pull request if you really need this feature.
